packages/dx-react-grid/docs/guides/custom-plugin-development.md

```diff
@@ -0,0 +1,708 @@
+
+```jsx
+...
+import {
+  ...,
+  TemplatePlaceholder
+} from '@devexpress/dx-react-core';
+
+export class ToolbarFilter extends React.PureComponent {
+  render() {
+    return (
+      <Plugin
+        ...
+      >
+        <Template name="toolbarContent">
+          <TemplatePlaceholder />
+        </Template>
+      </Plugin>
+    );
+  }
+}
+```
+
+Our plugin should not work without the [Toolbar](../reference/toolbar.md) plugin because we use its template. Specify the `dependencies` property to notify our plugin about it:
+
+```jsx
+...
+
+const pluginDependencies = [
+  { name: "Toolbar" }
+];
+
+export class ToolbarFilter extends React.PureComponent {
+  render() {
+    return (
+      <Plugin
+        ...
+        dependencies={pluginDependencies}
+      >
+        ...
+      </Plugin>
+    );
+  }
+}
+```
+
+Now, all is set to add custom markup to the `toolbarContent` template. Since we are going to display a select box, let's add the [Select](https://material-ui.com/demos/selects/) component to our template:
+
+```jsx
+...
+import Select from '@material-ui/core/Select';
+
+export class ToolbarFilter extends React.PureComponent {
+  render() {
+    return (
+      <Plugin
+        ...
+      >
+        <Template name="toolbarContent">
+          <TemplatePlaceholder />
+          <Select value="" />
+        </Template>
+      </Plugin>
+    );
+  }
+}
+```
+It is time to add our plugin to the Grid:
+
+*index.js:*
+```jsx
+...
+import { ToolbarFilter } from './plugins/toolbar-filter.js';
+
+class App extends Component {
+  ...
+  render() {
+    const { rows, columns } = this.state;
+    return (
+      <Paper>
+        <Grid
+          rows={rows}
+          columns={columns}
+        >
+          <Table />
+          <TableHeaderRow />
+          <Toolbar />
+          <ToolbarFilter />
+        </Grid>
+      </Paper>
+    );
+  }
+}
+```
+After you've done that, the Grid should look as follows:
+<p align="center">
+  <img class="img-responsive" src="../../../../../img/guides/custom-plugin-development/add-select.png">
+</p>
+
+## Share Values Between Plugins
+
+We see our select box, but it does nothing. The next step we should perform is to add items to our select box. These items should represent unique values of a particular column by which we want to filter our grid. The `ToolbarFilter` plugin represents a UI [plugin](./plugin-overview.md). This plugin should only render content based on provided data. But it should not manipulate the Grid's state or modify data. For this task, we need to implement an extra plugin called [state management](./plugin-overview.md) plugin. This plugin will prepare all necessary information for our `ToolbarFilter` plugin. Let's name it `ToolbarFilterState`. So, add a new `toolbar-filter-state.js` file to the `plugins` folder of our project and implement the logic of a new plugin in it.
```

Comment:
I would mention here that the UI/State plugin separation is optional. We can implement all stuff in a single plugin, but the separation of concerns make it more flexible and maintainable.